### DISCUSSION TITLE
failed to update dependencies: Internal server error

### DISCUSSION THREAD

**Discussion by bzarras** [deta/cloud-docs#658] [Help] [answered]
Hi, I'm not sure what I am doing wrong. I have a python fastapi project and want to add `deta` to my requirements.txt. Previously requirements.txt had just `fastapi` and I was able to deploy my app fine. Now with `deta` in there I am getting `failed to update dependencies: Internal server error`. I'm not sure if this is a problem on my side or if something is broken at Deta. Would appreciate any assistance!

My requirements.txt
```
deta==1.1.0
fastapi==0.88.0
```

What I get in the terminal
```
$ deta deploy
Updating dependencies...
Error: failed to update dependencies: Internal server error
```

Btw I am on macOS

**Comment by aavshr** [comment, marked as answer]
hi @bzarras sorry was an issue on our end, can you please try again now?

**Reply by bzarras** [reply]
Seems to be working now, thank you!

**Comment by begoon** [comment]
It seems I have a similar issue:
```
deta==1.1.0
fastapi==0.88.0
jinja2==3.1.2
passlib[bcrypt]==1.7.4
python-multipart==0.0.5
python-ulid==1.1.0
```

The deployment fails with:
```
Response:
{
	"errorMessage": "No module named 'fastapi'",
	"errorType": "ModuleNotFoundError",
	"requestId": "b826b9b1-4d17-4590-a405-c1798d4f3295",
	"stackTrace": [
		"  File \"/opt/python/detalib/debugger.py\", line 154, in wrap\n    raise e\n",
		"  File \"/opt/python/detalib/debugger.py\", line 142, in wrap\n    result = func(event, context)\n",
		"  File \"/var/task/_entry.py\", line 12, in handler\n    import main  # noqa\n",
		"  File \"/var/task/main.py\", line 9, in \u003cmodule\u003e\n    from fastapi import FastAPI, Form, Request, Response, UploadFile, status\n"
	]
}
```

I tried various fiddling, like clearing up `requirements.txt` and deploying to force the libraries to be deleted. Then I put the original `requirements.txt` back and deployed. Alas, no luck so far.

**Reply by begoon** [reply]
I also tried to delete the micro and deploy from scratch -- no luck.

Also, no luck with downgrading to python 3.8 and deploying that.

It seems that the dependencies installation is broken in general. The project works locally as expected.

**Reply by begoon** [reply]
https://github.com/deta/docs/discussions/654 looks like a similar issue.

**Comment by begoon** [comment]
I experimented more. I created a new micro from scratch. It is a trivial auto-generated python 3.9-based application, which `deta` CLI generates by default. It had only the `main.py` file but with no `requirements.txt`.

By default, when you don't instantiate the `app` variable explicitly with `Flask` or `FastAPI` constructors, the `app` variable is just a function.

The deployment of this micro worked with no problem. 

Then I added the `app = FastAPI()` line to use FastAPI. I also added the `fastapi` line to the newly created `requirements.txt`.

Now the deployment started failing to find the `fastapi` package. Ops!

Though according to the `.deta/state` file:
```
cat .deta/state
{"main.py":"9d8cca65862c5d60559b20a49dda44e489dc2fa3639c83b605115aa5fe0d2591","requirements.txt":"50b169668d64509a628d2b1a5e792ae5e19182844557a294e869e1797e150d2e"}
```

the `requirements.txt` file is there and deployed, but apparemently not used.

As it looks, the `deta` deployment tools do not process `requirements.txt` for some reason, causing the endpoint to fail.

Further, I added one more external dependency to the `requirements.txt` and used it in `main.py`. For example, `requests`.

After that, the deployment started to fail on the `requests` packages import.

```
Response:
{
	"errorMessage": "No module named 'requests'",
	"errorType": "ModuleNotFoundError",
	"requestId": "eeb4630a-01ef-498d-b16c-644e8f5579cd",
	"stackTrace": [
		"  File \"/opt/python/detalib/debugger.py\", line 131, in wrap\n    return func(event, context)\n",
		"  File \"/var/task/_entry.py\", line 12, in handler\n    import main  # noqa\n",
		"  File \"/var/task/main.py\", line 1, in \u003cmodule\u003e\n    import requests\n"
	]
}
```

So far, after all these steps, it looks that the `deta` deployment tools do not process the `requirements.txt` file at all, and this causes the deployments to fail.

**Reply by aavshr** [reply]
hi @begoon thanks for the detail investigation but I can not seem to replicate the error, do you get any logs or errors when you execute `deta deploy`?  You should get the installation logs when deploying, if not can you please check the following? 

- `requirements.txt` must be in the root directory of your Micro
- the `requirements.txt` file encoding should be `UTF-8` 

Can you please also check the output of the `deta details` command? If dependencies are installed, there should be list of dependencies in the output.

**Reply by begoon** [reply]
```
❯ deta deploy
Deploying...
Successfully deployed changes
Updating dependencies...
Error: failed to update dependencies: Internal server error

❌1 ❯ deta details
{
	"name": "bug",
	"id": "cbf3988c-1810-45eb-bec3-d27b919a5211",
	"project": "default",
	"runtime": "python3.9",
	"endpoint": "https://ze3qqy.deta.dev",
	"region": "eu-west-2",
	"visor": "enabled",
	"http_auth": "disabled"
}

❯ cat .deta/state
{"main.py":"2d0a413f2ae00b20ce22d258947608dfce71a99dd03afc70999c215f417a82ab","requirements.txt":"8c2b3b636159723d4cd159014afba238ac4fbc8046df5dc2a863421d236b93d6"}
```

**Reply by begoon** [reply]
`❯ cat main.py`
```
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def index():
    return {"ok": 1}
```

`❯ cat requirements.txt`
```
fastapi
```

**Reply by begoon** [reply]
`deta logs`

```
[2022-12-27T11:20:13Z] [ERROR] ModuleNotFoundError: No module named 'fastapi'
Traceback (most recent call last):
  File "/opt/python/detalib/debugger.py", line 154, in wrap
    raise e
  File "/opt/python/detalib/debugger.py", line 142, in wrap
    result = func(event, context)
  File "/var/task/_entry.py", line 12, in handler
    import main  # noqa
  File "/var/task/main.py", line 1, in <module>
    from fastapi import FastAPI
```

**Reply by aavshr** [reply]
> Error: failed to update dependencies: Internal server error

this was the culprit, it should work now sorry.

**Reply by aavshr** [reply]
can you please redeploy with `deta deploy` and try again?

**Reply by begoon** [reply]
Yes, now both micros (the original, where I started getting deployment issues, and the test one) now deploy and work as expected.

Thanks!

Seems, you tweaked something on the server side, and the issue is gone.
